### DISCUSSION TITLE
Burn Print Edition

### DISCUSSION THREAD

**Discussion by samuelvanderwaal** [metaplex-foundation/metaplex-program-library#689] [Ideas]
Metaplex is adding a `burn_edition_nft` handler that allows burning edition NFTs. The draft PR is [here](https://github.com/metaplex-foundation/metaplex-program-library/pull/684). This is a discussion for the community to ask questions about it, make suggestions, and generally have a say in this new feature.

This discussion will run for a week, until Tuesday, August 30th.
**EDIT: The discussion is extended a week until September 6th.**

The functionality currently implemented in the PR:

* Print editions passed in can be burned by the current owner of the print edition NFT--defined by the wallet with the token account holding the only token from the print edition NFT's mint account. 

* When a print edition is burned, the `supply` of the master edition is decremented. 

* Burning closes the token account, metadata account, edition account, and if applicable, the edition marker account

* The handler finds the edition's specific marker PDA and the position of the edition in the bit mask. After burning it flips the edition bit in the mask to zero. If all bits in the edition marker are now zero, then it also closes the edition marker PDA.

* All funds from closed accounts go to the edition owner/holder.

Some points of discussion:

Not decrementing the `max_supply` means that the update authority of the master edition original NFT can re-issue the burned print editions. If we were to instead decrement the `total_supply` every time a print edition was burned it would result in weird situations where you could have an edition numbered higher than the total supply: e.g. 13/8 edition. 

Closing the edition marker PDA when the last edition in its mask is burned means whoever the caller is gets that extra SOL. This is approximately 0.001 SOL, which is small enough that it shouldn't incentivize any weird gaming of burning print editions, but seems worth calling out explicitly.

For clarity, here is the Master Edition on-chain struct with the `supply` and `max_supply` fields:

```rust
pub struct MasterEditionV2 {
    pub key: Key,

    pub supply: u64,

    pub max_supply: Option<u64>,
}
```

`supply` represents the current number of prints that exist and the `max_supply` is the maximum allowed. If `max_supply` is `None` it means that there is no maximum and an infinite number of print editions can exist. If `max_supply` is `Some(0)`, it means that 0 print editions can exist, otherwise `Some(n)` means `n` print editions are allowed.

UPDATE:
I think we're converging on a spec so I'm going to write it here and will update it as needed.

**Final requirements and scope can be found [here](https://metaplex.notion.site/PRD-PUBLIC-Burn-Print-Edition-f9aae954fbc342e8996f92e1421874f9).**

**Comment by statikdev** [comment]
First of all, thanks @samuelvanderwaal for building this! I am going to benefit immensely from this.

> Print editions passed in can be burned by the owner of the NFT (as defined by the wallet with the token account holding the only token from the NFT's mint account).

For clarity, I assume you mean that the metaplex full burn of the edition can only be done by the wallet who owns the Master Edition? The rest of my thoughts are related so I will hold off on those.

**Reply by johnm** [reply]
Hm... that doesn't make sense. That would allow the Master Edition owner to destroy the print edition out from under the print edition's onwer.

**Reply by statikdev** [reply]
I read it as: "the print edition must be in the same wallet as the Master Edition". Meaning -- you would have to get the print edition back. However, thinking on it more, that makes little to no sense. 🤦

**Reply by samuelvanderwaal** [reply]
It's the owner/holder of the print edition itself, not of the master edition and not the update authority. I edited the description to try to make that more clear--let me know if it's still confusing. 

This is a good question though and prompted me to write a test specifically for this case.

**Comment by johnm** [comment]
Re: (not) decrementing the total supply 

Hm... It seems like there are case where either approach is the one desired. Two aspects pop into mind:
* what combinations of behaviors are curtailed by this proposed behavior? (Non-)decreasing x open/closed to further minting
* how will the behavior effect the use of those NFTs in e.g. DAO governance?

**Reply by samuelvanderwaal** [reply]
Does the governance module actually use print edition NFTs or are you thinking of sized collections? If print editions are actually used in DAO governance this is definitely something we need to take into account but my understanding was they used sized collections which relates to master edition NFTs so isn't affected by burning a print edition.

**Reply by johnm** [reply]
From the DAO creator/"operator"/"designer" perspective, we need the ability to support the use of any type of token in the operation of the DAO. The various types of tokens are created for a variety of different primary reasons and e.g., voting may be a secondary driver (so we don't want the "tail" of different token types to wag/"restrict" the dog).

In terms of technical aspects, the governance module supports various plugins. One of which allows us to dynamic specify e.g., the "max vote weight" for the given token type. That's an escape hatch which allows us to use non-fixed/managed-size collections in voting.

So, yes, I think it's critical to take the (potential) usage of Master+Print Editions in governance into account. Basically we should take governance into any/all potential changes to token types/policies.

**Reply by samuelvanderwaal** [reply]
The `supply` field can be used to indicate the "max vote weight", as it should accurately reflect "total number of print editions in existence". If people want a theoretical maximum number of prints, whether minted or not, they will be able to use the `max_supply` field to indicate how many prints can ever be created. 

If we allow decrementing the `max_supply` field, that will allow users to change the "max vote weight" by burning their prints, assuming the DAO is set up to use that field. 

If we disallow decrementing the `max_supply` field, users can't affect the "max vote weight" but the update authority of the Master Edition/Original NFT will be able to re-mint any burned prints and distribute them to other people. I'm not sure if this is desirable behavior for a DAO but it probably depends on the context.

**Reply by samuelvanderwaal** [reply]
I'm looking at `supply` in the code and remembering it's not even the actual current total supply.

**Reply by samuelvanderwaal** [reply]
The current implementation of `supply` is very counter-intuitive as it actually tracks the "highest edition number minted" not "current number of print editions minted" so it should not be relied on in governance for total weight stake. I'm proposing a fix for tracking supply going forward but I can't backfill all the current ones so I recommend against using Print Editions in governance.

**Comment by statikdev** [comment]
> Not decrementing the total supply means that the update authority of the master edition original NFT can re-issue the burned print editions.

I initially felt this was the best course of action but after giving it some thought, I'm wondering if we should modify it a bit.

If a user (not the update authority of the NFT) burns the edition, another edition should not be mintable -- meaning the total supply is `Y` but now, only `Y - 1` will ever exist.

If the update authority owns the print edition and burns it, the UA should be able to reprint the edition. This would allow for use cases where the master edition had bad data and needs to reprint to fix.

> If we were to instead decrement the total supply every time a print edition was burned it would result in weird situations where you could have an edition numbered higher than the total supply: e.g. 13/8 edition.

I agree this should be avoided.

**Reply by samuelvanderwaal** [reply]
> If a user (not the update authority of the NFT) burns the edition, another edition should not be mintable -- meaning the total supply is Y but now, only Y - 1 will ever exist.

Not sure I follow this. You want the total supply to be decremented when a user burns but not the update authority? In that case the supply would be Y - 1, right?

> > If we were to instead decrement the total supply every time a print edition was burned it would result in weird situations where you could have an edition numbered higher than the total supply: e.g. 13/8 edition.

> I agree this should be avoided.

Any suggestions on how to avoid this while decrementing the total supply? I don't think we can force users to burn in order.

**Reply by samuelvanderwaal** [reply]
Also, see my updates in the main post with clarifications on what `supply` and `max_supply` are.

**Reply by statikdev** [reply]
> Not sure I follow this. You want the total supply to be decremented when a user burns but not the update authority? In that case the supply would be Y - 1, right?

I think there are two possible events here:

**Print Edition Burned By Owner**

An artist creates 10 print editions of their work. Print edition `#7` is burned by the current owner (thrown away) who does not have `update_authority` on the print and master edition.

In this instance, I feel the `supply` and `max_supply` should stay the same. The owner made the choice to destroy the copy and therefore even though there is one less edition in the wild, `#7` is now essentially lost to time. The artist shouldn't be able to reprint because it was sold at one point and could lead to odd mechanisms where they sell `#7`, have someone burn it, sell again, etc.

**Print Edition Burned by Artist**

An artist creates 10 print editions of their work. Print edition `#7` is burned by the artist who has `update_authority` on the print and master edition.

In this instance, I feel the `supply` should decrement by 1 and the `max_supply` should still be 10. The artist made the choice to destroy the copy because they might want to reprint it or make another change.

I've never fully grokked the `is_primary_sale_happened`, but maybe if this isn't set, then it can be burned and reissued?

> If we were to instead decrement the total supply every time a print edition was burned it would result in weird situations where you could have an edition numbered higher than the total supply: e.g. 13/8 edition.
>
> Any suggestions on how to avoid this while decrementing the total supply? I don't think we can force users to burn in order.

Just to be clear, I am assuming `total supply` = `max_supply`. I believe we should never change the `max_supply` when a burn occurs. If the artists sets a fixed supply of 100 and feels they only want 50 to exist before they release, I feel they should have to burn it all and do it again.

Or -- is it currently possible to update the `max_supply` down to the current `supply` if the artist doesn't plan to print any more editions? I could see the artist wanting to update the `max_supply` if they have it up on a marketplace to print and it doesn't "print" out.

I can't think of other situations where we would reduce `max_supply`.  

[edit 1: wrap `#7` in backticks so it doesn't link to issues...]
[edit 2: include example for not printing out]

**Reply by statikdev** [reply]
Might be useful to poll some artists who use editions and get their feedback?

**Reply by samuelvanderwaal** [reply]
I was talking to Sammy about this and he pointed out that the way to prevent re-prints is to not flip the bit in the bit mask when burning rather than worrying about decrementing the `max_supply` number. That also gives us more flexibility if we decide to e.g. implement your suggestion here where update authority can reprint as we can determine whether to flip the bit by checking if owner == update authority.

> Print Edition Burned By Owner
> 
> An artist creates 10 print editions of their work. Print edition #7 is burned by the current owner (thrown away) who does not have update_authority on the print and master edition.
> 
> In this instance, I feel the supply and max_supply should stay the same. The owner made the choice to destroy the copy and therefore even though there is one less edition in the wild, #7 is now essentially lost to time. The artist shouldn't be able to reprint because it was sold at one point and could lead to odd mechanisms where they sell #7, have someone burn it, sell again, etc.
> 

In this case, if we use the bit mask to determine re-print eligibility, I think we'd want to still decrement the supply here for consistency, as ideally want supply to track current supply.

**Reply by samuelvanderwaal** [reply]
>I've never fully grokked the is_primary_sale_happened, but maybe if this isn't set, then it can be burned and reissued?

`primary_sale_happened` is used for "should we charge royalties yet or not?". 

Primary sale happened == false, don't charge royalties
Primary sale happened == true, charge royalties

With e.g. Candy Machine, you set it to true when minting the NFT because you want the initial mint price to go to the treasury wallet, but any sales thereafter should apply royalty. 

When minting e.g. a 1/1 and selling on a storefront, you leave the primary sale happened to false until after it's sold on the store.

**Reply by statikdev** [reply]
> I was talking to Sammy about this and he pointed out that the way to prevent re-prints is to not flip the bit in the bit mask when burning rather than worrying about decrementing the max_supply number. That also gives us more flexibility if we decide to e.g. implement your suggestion here where update authority can reprint as we can determine whether to flip the bit by checking if owner == update authority.

Makes sense! 

> In this case, if we use the bit mask to determine re-print eligibility, I think we'd want to still decrement the supply here for consistency, as ideally want supply to track current supply.

Just to be clear, do you meant decrement the supply whether we allow reprints or not?

**Reply by samuelvanderwaal** [reply]
> Just to be clear, do you meant decrement the supply whether we allow reprints or not?

Correct.
